FIX: if bins input to hist is str, treat like no bins

tacaswell opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
https://matplotlib.org/devdocs/devel/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the
PR out of master, but out of a separate branch.  See
https://matplotlib.org/devel/gitwash/development_workflow.html for
instructions.-->

## PR Summary
This change causes the range of all data sets to be computed and
passed to numpy (which in turn uses the total range to compute the
'best' bins).

The existing code 'latches' the bins from the first data set to use
for the rest so this is can still lead to poor binning (if the data
sets are widely different).

closes #8636

This still needs some tests and possibly a more through thinking of how to deal with the 'string' bins and multiple data sets.

Computing the histogram once with all of the data and the using those bins may be a better option (but involves computing the histograms twice).


<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

anntzer commented:
> Computing the histogram once with all of the data and the using those bins may be a better option (but involves computing the histograms twice).

... unless you're willing to lookup the bin selector into the private `np.lib.function_base._hist_bin_selectors`.

tacaswell commented:
Even if you are willing to lookup the selector, you still have to do the histogram once with all of the data and once for each data set.

anntzer commented:
No, you only want to compute the bins once on the total dataset, and then histogram each individual dataset using the given bins, right?

eric-wieser commented:
> Computing the histogram once with all of the data and the using those bins may be a better option (but involves computing the histograms twice).

I think that this might be an acceptable cost. I'm hoping to get things fixed in numpy in 1.15, but it'll be a while before matplotlib's minimum version can rely on it.

Wouldn't it be better to have a 2x slowdown in exchange for correctness?

I suppose you could backport the histrogram functions from 1.15.

tacaswell commented:
Ah, I now understand what @anntzer is saying about using the selector.  I am pretty :-1: on touching numpy internals and checking for something that does not exist in all versions of numpy that we support.  Open to vendoring the selectors (despite them being in numpy because we bounced a feature request from Matplotlib to numpy), but probably not the full histogram code (we just about finished getting rid of a partial backport of histogram!).

eric-wieser commented:
> I am pretty 👎 on touching numpy internals

As you should be, because they're about to move

> Open to vendoring the selectors

If 1.15 adds `histogram_edges`, then you could vendor that function

anntzer commented:
numpy 1.15 is out, now we just need to bump the dependency to it :)

tacaswell commented:
rebased and added a test, still not sure if this is the right thing.

On one hand, it fixes a problem (when people pass in more than on distribution and bins as a string it was running the bin generation algorithm only considering the first dataset's limits) but only part way (it now uses the correct limits, but only considers the distribution of the first dataset).

eric-wieser commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6637,7 +6637,9 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
-        binsgiven = np.iterable(bins) or bin_range is not None
+        binsgiven = ((np.iterable(bins) and
+                      not isinstance(bins, str)) or
+                     bin_range is not None)
```
It might be worth matching the numpy logic here, as:
```python
if bins is None:
    binsgiven = False
else:
    # match the logic in numpy.lib.histograms._get_bin_edges
    if isinstance(bins, basestring):
        binsgiven = False
    elif np.ndim(bins) == 1:
        binsgiven = True
    else:
        binsgiven = False
```
Renaming to `bins_array_given` would make sense too.

eric-wieser commented:
Would something like this be a better compromise?

```python
try:
    from numpy.lib.histograms import histogram_bin_edges
except ImportError:
    def histogram_bin_edges(arr, bins, range=None, weights=None):
        if isinstance(bins, basestring):
            # rather than backporting the internals, just do the full computation.
            # If this is too slow for users, they can update numpy, or pick a manual number of bins
            return np.histogram(arr, bins, range, weights)[1]
        elif np.ndim(bins) == 0:
            # not strictly the behavior of histogram_bin_edges, but easier to not bother with the linspace here
            return bins
        else:
            return bins
````

tacaswell commented:
> bins = histogram_bin_edges(arr.ravel(), bins, range, weights)

Except that we are not guaranteed that the input is a 2D numpy array (or even equal length). Would probably have to do something like `hstack` to be safe....

Applied the in-line comments gouh.

eric-wieser commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6666,7 +6675,7 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
         # we need to figure out the range required for all datasets,
         # and supply that to np.histogram.
-        if not binsgiven and not input_empty:
+        if not bins_array_given and not input_empty:
```
Think you now need `and bin_range is None` here, which makes more sense anyway

eric-wieser commented:
> Would probably have to do something like `hstack` to be safe

I think `concatenate` would be more appropriate here. Was hoping for an approach that wouldn't make extra copies, but I'm not sure if that's a worthwhile or avoidable concern anyway.

eric-wieser commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6063,3 +6063,9 @@ def invert(x):
+
+
+def test_hist_auto_bins():
+    _, bins, _ = plt.hist([[1, 2, 3], [4, 5, 6]], bins='auto')
```
A test on jagged input might be nice too

eric-wieser commented:
PR description is pretty outdated after the title change

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6637,7 +6637,16 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
 
         # Check whether bins or range are given explicitly.
-        binsgiven = np.iterable(bins) or bin_range is not None
+        if bins is None:
```
The whole thing is just `bins_array_given = np.ndim(bins) == 1`? (perhaps with additional comments indicating that this excludes str and None).

anntzer commented:
I agree with concatenating everything and passing down to numpy.  I doubt that's even close to being a bottleneck (and anyways users can pass in their own bins if they really care).

eric-wieser commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6666,19 +6681,19 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
-                if len(xi) > 0:
-                    xmin = min(xmin, np.nanmin(xi))
-                    xmax = max(xmax, np.nanmax(xi))
-            bin_range = (xmin, xmax)
```
You either need to keep this code for the case when `bins` is an integer, or you need to add a case for integral bins in the `histogram_bin_edges` backport

tacaswell replied:
ah, good catch on that, I fell down a different rabbit hole (see my comment at the issue level)...

tacaswell commented:
so, I think this found a bug / numerical instability in numpy?!

```python
import numpy as np
from numpy.lib.histograms import histogram_bin_edges

print(np.__version__)

np.random.seed(0)
data = np.random.standard_normal(2000)
data += -2.0 - np.min(data)
data_pos = data + 2.1
data_big = data_pos + 30
weights = np.ones_like(data) * 1.e-5


m, b = np.histogram(data, 100, weights=weights)
m2, b2 = np.histogram(data, b, weights=weights)

bc = histogram_bin_edges(data, 100, None, weights)
m3, b3 = np.histogram(data, bc, weights=weights)

print(m == m2)
print(m == m3)
print(m2 == m3)
print(m - m2)

```

on ~numpy master 

you get things like

```
jupiter@23:33  ➤  python /tmp/np_test.py 
1.17.0.dev0+1bf2b94
[ True  True  True  True  True  True False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False  True  True
  True  True  True False]
[ True  True  True  True  True  True False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False  True  True
  True  True  True False]
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True]
[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
  0.00000000e+00  0.00000000e+00  3.38813179e-21  6.77626358e-21
 -1.01643954e-20  1.69406589e-21  3.38813179e-21  1.69406589e-21
  6.77626358e-21  6.77626358e-21 -1.01643954e-19 -7.45388994e-20
 -1.55854062e-19 -2.03287907e-19 -5.08219768e-20 -2.57498016e-19
 -3.25260652e-19 -2.84603070e-19 -5.42101086e-19 -4.06575815e-19
 -3.25260652e-19 -5.42101086e-19 -5.14996032e-19 -5.14996032e-19
 -1.62630326e-19 -5.69206141e-19 -6.50521303e-19 -6.50521303e-19
 -6.50521303e-19 -6.50521303e-19 -6.50521303e-19  1.40946282e-17
  1.75640752e-17  1.49619900e-17  1.79977561e-17  1.36609474e-17
  1.79977561e-17  1.97324795e-17  2.14672030e-17  2.14672030e-17
  2.01661604e-17  1.62630326e-17  1.97324795e-17  2.14672030e-17
  2.23345648e-17  2.84060969e-17  2.01661604e-17  1.97324795e-17
  2.58040117e-17  1.79977561e-17  1.88651178e-17  2.32019265e-17
  1.88651178e-17  1.97324795e-17  1.53956708e-17  1.27935856e-17
  1.06251813e-17  1.40946282e-17  1.40946282e-17  1.45283091e-17
  1.62630326e-17  1.19262239e-17  9.37834879e-18  7.75204553e-18
  5.31259065e-18  7.34546972e-18  8.97177298e-18  8.97177298e-18
  6.12574227e-18  6.12574227e-18  5.31259065e-18  4.07931067e-18
  6.12574227e-18  4.48588649e-18  2.85958323e-18  3.67273486e-18
  3.26615904e-18  2.85958323e-18  4.08269881e-19  2.85958323e-18
  2.03965534e-18  2.44623115e-18  1.63307952e-18  1.22650371e-18
  4.08269881e-19  1.63307952e-18  4.08269881e-19  8.16539761e-19
  8.16539761e-19  4.08269881e-19  0.00000000e+00  0.00000000e+00
  0.00000000e+00  0.00000000e+00  0.00000000e+00  4.08269881e-19]

```

This seems to be enough to cause the y limits on one of the histogram tests to break...

eric-wieser commented:
What you're seeing there is that numpy has an optimization for when the bins are uniform, and like many optimizations it leads to incorrect results. That optimization has bitten us repeatedly...

tacaswell commented:
ok, fixed the failing test by only going through this logic iff we have more than 1 input, and I think I addressed the issue about integer input that @eric-wieser brought up.

eric-wieser commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6636,8 +6655,8 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
-        # Check whether bins or range are given explicitly.
-        binsgiven = np.iterable(bins) or bin_range is not None
+        # this in True for 1D arrays, and False for None and str
+        bins_array_given = np.ndim(bins) == 1
```
At this point you may as well just put this inside `histogram_bin_edges`, since it's there in the non-backport

tacaswell commented:
On closer inspection, this was a real failure.  Fixed locally...

eric-wieser commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -36,6 +36,25 @@
+                # hard-code numpy's default
+                bins = 10
+            if range is None:
+                range = np.nanmin(arr), np.manmax(arr)
```
Note that numpy does not use the nan variants here.

anntzer replied:
Agreed that if we want to support nans, we should pre-remove them ourselves before passing them to histogram_bin_edges.

jklymak replied:
... also not the typo here....

tacaswell replied:
Just took me about 15 minutes to sort out what the typo was....

tacaswell commented:
The test failure is the svg fix from #13710 .

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6638,22 +6657,38 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
                     xmin = min(xmin, np.nanmin(xi))
                     xmax = max(xmax, np.nanmax(xi))
-            bin_range = (xmin, xmax)
+            if xmax > xmin:
```
looks like this line is only to handle the case where all x's are the same?  but is that really needed? (why was it not there before?)

tacaswell replied:
was thinking about the all-nan case, should probably be `>=`?

tacaswell replied:
pushed a change for this

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6664,7 +6699,7 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
             m, bins = np.histogram(x[i], bins, weights=w[i], **hist_kwargs)
             m = m.astype(float)  # causes problems later if it's an int
-            if mlast is None:
+            if stacked and mlast is None:
```
is that fixing a separate bug (or not?)

tacaswell replied:
mirco-optomization, we only care if this is non-None if we are stacking.

anntzer replied:
then move the condition below into this one
```
if stacked:
    if mlast is None: mlast = ...
    m += mlast
    mlast[:] = m
```
or even better
```
mlast = 0  # instead of None
if stacked:
    m += mlast
    mlast = m.copy()
```
seems easier to follow (sure, you don't get to reuse the same buffer, but I doubt that's relevant).

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -36,6 +36,25 @@
+try:
+    from numpy.lib.histograms import histogram_bin_edges
+except ImportError:
+    def histogram_bin_edges(arr, bins, range=None, weights=None):
```
add a comment mentioning which version of numpy this was introduced in (so that we can remove the backport when possible)

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6638,22 +6657,40 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
                     xmax = max(xmax, np.nanmax(xi))
-            bin_range = (xmin, xmax)
+            # make sure we have at least one non-nan value
+            # before we
```
"before we"... what?

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6638,22 +6657,40 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
+        # if the bin_range is not given, compute without nan numpy
+        # does not do this for us when guessing the range (but will
+        # happily ignore nans when computing the histogram).
+        if bin_range is None:
```
just `xmin = min((np.nanmin(xi) for xi in x if xi.size), default=np.inf)` and the same for xmax?  seems clearer

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6638,22 +6657,40 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
             xmax = -np.inf
             for xi in x:
-                if len(xi) > 0:
+                if len(xi):
```
so we ignore nans, but not infs? (dunno, just making sure)

tacaswell replied:
This maintains the current behavior of 3.0

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6638,22 +6657,40 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
+        # If bins are not specified either explicitly or via range,
+        # we need to figure out the range required for all datasets,
+        # and supply that to np.histogram.
+        if not bins_array_given and not input_empty and len(x) > 1:
```
the checks for input_empty and len(x) > 1 are redundant I think?

tacaswell replied:
The `len(x) > 1` is to avoid triggering the numerical issue in numpy described in https://github.com/matplotlib/matplotlib/pull/8638#issuecomment-467292466.

These are also testing slightly different things as `np.array([[], []])` fails the first one, but not the second.

tacaswell commented:
Squashed this down to 4 commits.

Given the way this is going it is not going to backport cleanly....
